lib/task/digitalobject/digitalObjectDeleteTask.class.php

```diff
@@ -50,71 +53,132 @@ protected function configure()
     {
-      throw new sfException('Failed to fetch information object with the slug given.');
+      throw new sfException('Invalid slug with object type "'.$row['class_name'].'" entered.');
     }
 
-    $ios = $options['and-descendants'] ? $infoObj->descendants->andSelf() : array($infoObj);
-    $nDeleted = 0;
+    if ($row['class_name'] == 'QubitInformationObject' &&
+      null === $informationObject = QubitInformationObject::getById($row['object_id']))
+    {
+      throw new sfException('Failed to fetch information object with the slug given.');
+    }
+    else if ($row['class_name'] == 'QubitRepository' &&
+      null === $repository = QubitRepository::getById($row['object_id']))
+    {
+      throw new sfException('Failed to fetch repository with the slug given.');
+    }
 
-    foreach ($ios as $io)
+    switch ($row['class_name'])
     {
-      $this->logSection('digital-object', sprintf('(%d of %d) deleting digital object for: %s',
-                        ++$nDeleted, count($ios), $io->getTitle(array('cultureFallback' => true))));
+      case 'QubitInformationObject':
+        $objectIds = $options['and-descendants'] ? $this->getInformationObjectChildren($informationObject) : array($informationObject->id);
+
+        break;
+
+      case 'QubitRepository':
+        // Get all linked top level information object recs.
+        $relatedInformationObjects = $repository->getRelatedInformationObjectIds();
+
+        foreach ($relatedInformationObjects as $id)
+        {
+          $object = QubitInformationObject::getById($id);
+
+          // Always include descendants when deleting by repository.
+          $objectIds = array_merge($objectIds, $this->getInformationObjectChildren($object));
+        }
+
+        break;
+    }
 
-      // Remove appropriate digital object files, empty directories left behind, and db entries
-      foreach ($io->digitalObjectsRelatedByobjectId as $do)
+    foreach ($objectIds as $id)
+    {
+      if (null !== $object = QubitInformationObject::getById($id))
       {
-        $this->deleteDigitalObjectFiles($do);
-        QubitDigitalObject::pruneEmptyDirs(dirname($do->getAbsolutePath()));
-        $do->delete();
+        $success = $this->deleteDigitalObject($object, $options);
+
+        $this->logSection('digital-object', sprintf('(%d of %d) %s: %s',
+          $success ? ++$nDeleted : $nDeleted,
+          count($objectIds),
+          $success ? "deleting digital object for" : "nothing to delete",
+          $object->getTitle(array('cultureFallback' => true)))
+        );
       }
+      Qubit::clearClassCaches();
     }
 
-    $this->logSection('digital-object', sprintf('%d digital objects deleted successfully (%.2fs elapsed)', $nDeleted, $t->elapsed()));
+    $this->logSection('digital-object', sprintf('%d digital objects deleted (%.2fs elapsed)', $nDeleted, $t->elapsed()));
   }
 
-  private function deleteDigitalObjectFiles($digitalObject)
+  private function getInformationObjectChildren($object)
```

Comment:
I'd call it `getIoDescendantIds` and use the `lft` and `rgt` values as parameters directly, as mentioned above.

Reply:
See above. This part was changed to accept lft and rgt values.